### ISSUE TITLE
Some boolean settings are displayed as strings

### ISSUE THREAD

**Issue by deivid-rodriguez** [bundler/bundler#6672]
Yesterday I turned on "bundler 2 mode" by dropping the following file in `~/.bundle/config`

```yaml
---

BUNDLE_GEM__TEST: "rspec"
BUNDLE_GEM__MIT: "true"
BUNDLE_GEM__COC: "true"
BUNDLE_ALLOW_BUNDLER_DEPENDENCY_CONFLICTS: "true"
BUNDLE_ALLOW_OFFLINE_INSTALL: "true"
BUNDLE_AUTO_CLEAN_WITHOUT_PATH: "true"
BUNDLE_AUTO_CONFIG_JOBS: "true"
BUNDLE_CACHE_ALL: "true"
BUNDLE_CACHE_COMMAND_IS_PACKAGE: "true"
BUNDLE_CONSOLE_COMMAND: "true"
BUNDLE_DEFAULT_INSTALL_USES_PATH: "true"
BUNDLE_DEPLOYMENT_MEANS_FROZEN: "true"
BUNDLE_DISABLE_MULTISOURCE: "true"
BUNDLE_ERROR_ON_STDERR: "true"
BUNDLE_FORGET_CLI_OPTIONS: "true"
BUNDLE_GLOBAL_PATH_APPENDS_RUBY_SCOPE: "true"
BUNDLE_GLOBAL_GEM_CACHE: "true"
BUNDLE_INIT_GEMS_RB: "true"
BUNDLE_LIST_COMMAND: "true"
BUNDLE_LOCKFILE_USES_SEPARATE_RUBYGEMS_SOURCES: "true"
BUNDLE_ONLY_UPDATE_TO_NEWER_VERSIONS: "true"
BUNDLE_PATH_RELATIVE_TO_CWD: "true"
BUNDLE_PLUGINS: "true"
BUNDLE_PREFER_GEMS_RB: "true"
BUNDLE_PRINT_ONLY_VERSION_NUMBER: "true"
BUNDLE_SETUP_MAKES_KERNEL_GEM_PUBLIC: "false"
BUNDLE_SKIP_DEFAULT_GIT_SOURCES: "true"
BUNDLE_SPECIFIC_PLATFORM: "true"
BUNDLE_SUPPRESS_INSTALL_USING_MESSAGES: "true"
BUNDLE_UNLOCK_SOURCE_UNLOCKS_SPEC: "true"
BUNDLE_UPDATE_REQUIRES_ALL_FLAG: "true"
BUNDLE_USE_GEM_VERSION_PROMOTER_FOR_MAJOR_UPDATES: "true"
BUNDLE_VIZ_COMMAND: "false"
BUNDLE_DEFAULT_CLI_COMMAND: "cli_help"
```

Everything is working as expected so far, but I see something "weird" when I run `bundle config`. This is the output I get:

```console
$ bundle config
Settings are listed in order of priority. The top value will be used.
gem.test
Set for the current user (/home/deivid/.bundle/config): "rspec"

gem.mit
Set for the current user (/home/deivid/.bundle/config): true

gem.coc
Set for the current user (/home/deivid/.bundle/config): true

allow_bundler_dependency_conflicts
Set for the current user (/home/deivid/.bundle/config): true

allow_offline_install
Set for the current user (/home/deivid/.bundle/config): true

auto_clean_without_path
Set for the current user (/home/deivid/.bundle/config): true

auto_config_jobs
Set for the current user (/home/deivid/.bundle/config): "true"

cache_all
Set for the current user (/home/deivid/.bundle/config): true

cache_command_is_package
Set for the current user (/home/deivid/.bundle/config): true

console_command
Set for the current user (/home/deivid/.bundle/config): true

default_install_uses_path
Set for the current user (/home/deivid/.bundle/config): true

deployment_means_frozen
Set for the current user (/home/deivid/.bundle/config): true

disable_multisource
Set for the current user (/home/deivid/.bundle/config): true

error_on_stderr
Set for the current user (/home/deivid/.bundle/config): true

forget_cli_options
Set for the current user (/home/deivid/.bundle/config): true

global_path_appends_ruby_scope
Set for the current user (/home/deivid/.bundle/config): "true"

global_gem_cache
Set for the current user (/home/deivid/.bundle/config): true

init_gems_rb
Set for the current user (/home/deivid/.bundle/config): true

list_command
Set for the current user (/home/deivid/.bundle/config): true

lockfile_uses_separate_rubygems_sources
Set for the current user (/home/deivid/.bundle/config): true

only_update_to_newer_versions
Set for the current user (/home/deivid/.bundle/config): true

path_relative_to_cwd
Set for the current user (/home/deivid/.bundle/config): "true"

plugins
Set for the current user (/home/deivid/.bundle/config): true

prefer_gems_rb
Set for the current user (/home/deivid/.bundle/config): true

print_only_version_number
Set for the current user (/home/deivid/.bundle/config): true

setup_makes_kernel_gem_public
Set for the current user (/home/deivid/.bundle/config): false

skip_default_git_sources
Set for the current user (/home/deivid/.bundle/config): true

specific_platform
Set for the current user (/home/deivid/.bundle/config): true

suppress_install_using_messages
Set for the current user (/home/deivid/.bundle/config): true

unlock_source_unlocks_spec
Set for the current user (/home/deivid/.bundle/config): true

update_requires_all_flag
Set for the current user (/home/deivid/.bundle/config): true

use_gem_version_promoter_for_major_updates
Set for the current user (/home/deivid/.bundle/config): "true"

viz_command
Set for the current user (/home/deivid/.bundle/config): false

default_cli_command
Set for the current user (/home/deivid/.bundle/config): "cli_help"

retry
Set for your local app (/home/deivid/Code/bundler/.bundle/config): 3

path
Set for your local app (/home/deivid/Code/bundler/.bundle/config): ".bundle"

jobs
Set for your local app (/home/deivid/Code/bundler/.bundle/config): "3"
```

As you can see, some of the boolean setting are printed as strings. Not sure if there's an underlying bug here, but it seems weird, right?

**Comment by segiddins** [comment]
Are they listed as Boolean settings in Bundler::Settings ?

**Comment by deivid-rodriguez** [comment]
Yeah.

Also seeing the same thing on numeric settings, `jobs`' value is printed as a string, whereas `retry`'s value as a number.

**Comment by segiddins** [comment]
I can't reproduce this on `master`:

```
Settings are listed in order of priority. The top value will be used.
gem.test
Set for the current user (bundle_config): "rspec"

gem.mit
Set for the current user (bundle_config): true

gem.coc
Set for the current user (bundle_config): true

allow_bundler_dependency_conflicts
Set for the current user (bundle_config): true

allow_offline_install
Set for the current user (bundle_config): true

auto_clean_without_path
Set for the current user (bundle_config): true

auto_config_jobs
Set for the current user (bundle_config): true

cache_all
Set for the current user (bundle_config): true

cache_command_is_package
Set for the current user (bundle_config): true

console_command
Set for the current user (bundle_config): true

default_install_uses_path
Set for the current user (bundle_config): true

deployment_means_frozen
Set for the current user (bundle_config): true

disable_multisource
Set for the current user (bundle_config): true

error_on_stderr
Set for the current user (bundle_config): true

forget_cli_options
Set for the current user (bundle_config): true

global_path_appends_ruby_scope
Set for the current user (bundle_config): true

global_gem_cache
Set for the current user (bundle_config): true

init_gems_rb
Set for the current user (bundle_config): true

list_command
Set for the current user (bundle_config): true

lockfile_uses_separate_rubygems_sources
Set for the current user (bundle_config): true

only_update_to_newer_versions
Set for the current user (bundle_config): true

path_relative_to_cwd
Set for the current user (bundle_config): true

plugins
Set for the current user (bundle_config): true

prefer_gems_rb
Set for the current user (bundle_config): true

print_only_version_number
Set for the current user (bundle_config): true

setup_makes_kernel_gem_public
Set for the current user (bundle_config): false

skip_default_git_sources
Set for the current user (bundle_config): true

specific_platform
Set for the current user (bundle_config): true

suppress_install_using_messages
Set for the current user (bundle_config): true

unlock_source_unlocks_spec
Set for the current user (bundle_config): true

update_requires_all_flag
Set for the current user (bundle_config): true

use_gem_version_promoter_for_major_updates
Set for the current user (bundle_config): true

viz_command
Set for your local app (/Users/segiddins/Desktop/GemfileTest/.bundle/config): true
Set for the current user (bundle_config): false

default_cli_command
Set for your local app (/Users/segiddins/Desktop/GemfileTest/.bundle/config): "cli_help"
Set for the current user (bundle_config): "cli_help"

retry
Set for your local app (/Users/segiddins/Desktop/GemfileTest/.bundle/config): 5

jobs
Set for your local app (/Users/segiddins/Desktop/GemfileTest/.bundle/config): 3

config
Set via BUNDLE_CONFIG: "bundle_config"
```

**Comment by deivid-rodriguez** [comment]
Oh, yeah, forgot to say this is on 1.16.4.

**Comment by segiddins** [comment]
I think we can close this as being fixed then? `bundle config` was rewritten shortly after 1.16 branched

**Comment by deivid-rodriguez** [comment]
Fine with me, since there's no specific misbehavior associated to problem, and it's fixed in master as you said :+1:

**Comment by deivid-rodriguez** [comment]
Ok, so I ended up getting to the bottom of this. The settings that are printed as strings are the only settings not available in 1.16-stable. Since bundler has no hint about them being boolean, it just prints them as strings.

This leads to 2 questions:

* Should this settings be backported so people can start opting in?
* Bundler currently gives no hint about unknown settings (like these four), so if you for example make a typo on a setting, it will go unnoticed. Should it gives a warning or error, should these warning/error behavior be configurable under a new setting?

**Comment by segiddins** [comment]
Non-string Settings aren’t enumerated anywhere, so there’s no way to know (at present) which are expected
